File: android/app/src/main/java/com/scholarlyapps/pathlingo/data/networking/ApiService.java
package com.scholarlyapps.pathlingo.data.networking;

import com.scholarlyapps.pathlingo.data.remote.dto.AppUserDto;
import com.scholarlyapps.pathlingo.data.remote.dto.AvatarDto;
import com.scholarlyapps.pathlingo.data.remote.dto.AuthResponse;
import com.scholarlyapps.pathlingo.data.remote.dto.CategoryDto;
import com.scholarlyapps.pathlingo.data.remote.dto.CoinsBalanceResponse;
import com.scholarlyapps.pathlingo.data.remote.dto.ForgotPasswordRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.GoogleLoginRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.GuestLoginRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.ListResponse;
import com.scholarlyapps.pathlingo.data.remote.dto.LoginRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.MessageResponse;
import com.scholarlyapps.pathlingo.data.remote.dto.OtpSendRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.OtpVerifyRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.RegisterRequest;
import com.scholarlyapps.pathlingo.data.remote.dto.SubcategoryDto;
import com.scholarlyapps.pathlingo.data.remote.dto.VocabularyDto;
import com.scholarlyapps.pathlingo.data.remote.dto.WrappedResponse;

import java.util.Map;

import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.PartMap;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface ApiService {

    @GET("categories")
    Call<ListResponse<CategoryDto>> getCategories();

    @GET("subcategories")
    Call<ListResponse<SubcategoryDto>> getSubcategories(@Query("category_id") Long categoryId);

    @GET("vocabularies")
    Call<ListResponse<VocabularyDto>> getVocabularies(
        @Query("search") String search,
        @Query("category_id") Long categoryId,
        @Query("subcategory_id") Long subcategoryId
    );

    @POST("auth/guest")
    Call<AuthResponse> guestLogin(@Body GuestLoginRequest body);

    @POST("auth/register")
    Call<AuthResponse> register(@Body RegisterRequest body);

    @POST("auth/login")
    Call<AuthResponse> login(@Body LoginRequest body);

    @POST("auth/social/google")
    Call<AuthResponse> loginWithGoogle(@Body GoogleLoginRequest body);

    @POST("auth/otp/send")
    Call<MessageResponse> sendOtp(@Body OtpSendRequest body);

    @POST("auth/otp/verify")
    Call<MessageResponse> verifyOtp(@Body OtpVerifyRequest body);

    @POST("auth/forgot-password")
    Call<MessageResponse> forgotPassword(@Body ForgotPasswordRequest body);

    @POST("auth/logout")
    Call<MessageResponse> logout();

    @GET("me")
    Call<WrappedResponse<AppUserDto>> me();

    @Multipart
    @POST("me")
    Call<WrappedResponse<AppUserDto>> updateProfile(@PartMap Map<String, RequestBody> fields);

    @GET("favorites")
    Call<Map<String, Object>> favorites();

    @POST("favorites/{vocabulary}")
    Call<MessageResponse> addFavorite(@Path("vocabulary") long vocabularyId);

    @DELETE("favorites/{vocabulary}")
    Call<MessageResponse> removeFavorite(@Path("vocabulary") long vocabularyId);

    @GET("me/reads")
    Call<Map<String, Object>> myReads();

    @POST("words/{vocabulary}/read")
    Call<Map<String, Object>> markRead(@Path("vocabulary") long vocabularyId);

    @GET("me/coins")
    Call<CoinsBalanceResponse> coinsBalance();

    @GET("me/coins/transactions")
    Call<Map<String, Object>> coinsTransactions();

    @GET("avatars")
    Call<ListResponse<AvatarDto>> getAvatars();

    @POST("avatars/{id}/purchase")
    Call<WrappedResponse<CoinsBalanceResponse>> purchaseAvatar(@Path("id") long avatarId);

    @PATCH("me/avatar")
    Call<WrappedResponse<AppUserDto>> setActiveAvatar(@Body Map<String, Long> body);

    @POST("words/{id}/unlock")
    Call<MessageResponse> unlockWord(@Path("id") long wordId);

    @POST("categories/{id}/unlock")
    Call<MessageResponse> unlockCategory(@Path("id") long categoryId);

    @POST("subcategories/{id}/unlock")
    Call<MessageResponse> unlockSubcategory(@Path("id") long subcategoryId);
}

File: android/app/src/main/java/com/scholarlyapps/pathlingo/data/remote/ApiService.kt
package com.scholarlyapps.pathlingo.data.remote

import com.scholarlyapps.pathlingo.data.remote.dto.AppUserDto
import com.scholarlyapps.pathlingo.data.remote.dto.AuthResponse
import com.scholarlyapps.pathlingo.data.remote.dto.CategoryDto
import com.scholarlyapps.pathlingo.data.remote.dto.CoinsBalanceResponse
import com.scholarlyapps.pathlingo.data.remote.dto.ForgotPasswordRequest
import com.scholarlyapps.pathlingo.data.remote.dto.GoogleLoginRequest
import com.scholarlyapps.pathlingo.data.remote.dto.ListResponse
import com.scholarlyapps.pathlingo.data.remote.dto.LoginRequest
import com.scholarlyapps.pathlingo.data.remote.dto.MessageResponse
import com.scholarlyapps.pathlingo.data.remote.dto.RegisterRequest
import com.scholarlyapps.pathlingo.data.remote.dto.SubcategoryDto
import com.scholarlyapps.pathlingo.data.remote.dto.VocabularyDto
import com.scholarlyapps.pathlingo.data.remote.dto.WrappedResponse
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface ApiService {

    @GET("categories")
    suspend fun getCategories(): ListResponse<CategoryDto>

    @GET("subcategories")
    suspend fun getSubcategories(
        @Query("category_id") categoryId: Long? = null,
    ): ListResponse<SubcategoryDto>

    @GET("vocabularies")
    suspend fun getVocabularies(
        @Query("search") search: String? = null,
        @Query("category_id") categoryId: Long? = null,
        @Query("subcategory_id") subcategoryId: Long? = null,
    ): ListResponse<VocabularyDto>

    @POST("auth/register")
    suspend fun register(@Body body: RegisterRequest): AuthResponse

    @POST("auth/login")
    suspend fun login(@Body body: LoginRequest): AuthResponse

    @POST("auth/social/google")
    suspend fun loginWithGoogle(@Body body: GoogleLoginRequest): AuthResponse

    @POST("auth/forgot-password")
    suspend fun forgotPassword(@Body body: ForgotPasswordRequest): MessageResponse

    @POST("auth/logout")
    suspend fun logout(): MessageResponse

    @GET("me")
    suspend fun me(): WrappedResponse<AppUserDto>

    @GET("favorites")
    suspend fun favorites(): Map<String, Any?>

    @POST("favorites/{vocabulary}")
    suspend fun addFavorite(@Path("vocabulary") vocabularyId: Long): MessageResponse

    @DELETE("favorites/{vocabulary}")
    suspend fun removeFavorite(@Path("vocabulary") vocabularyId: Long): MessageResponse

    @GET("me/reads")
    suspend fun myReads(): Map<String, Any?>

    @POST("words/{vocabulary}/read")
    suspend fun markRead(@Path("vocabulary") vocabularyId: Long): Map<String, Any?>

    @GET("me/coins")
    suspend fun coinsBalance(): CoinsBalanceResponse

    @GET("me/coins/transactions")
    suspend fun coinsTransactions(): Map<String, Any?>
}
